File: typescript/src/golf.test.ts
import { test, describe } from 'node:test';
import assert from 'node:assert/strict';
import * as fs from 'node:fs';
import * as path from 'node:path';
import { GolfWriter, type WriterConfig } from './writer.js';
import { GolfReader } from './reader.js';
import { Compression, TimestampResolution, type Record } from './core.js';

const TESTDATA_DIR = path.join(__dirname, '..', '..', 'testdata');

function makeValue(seed: number, size: number): Buffer {
  const buf = Buffer.alloc(size);
  for (let i = 0; i < size; i++) {
    buf[i] = (seed + i) & 0xff;
  }
  return buf;
}

function buildInMemory(
  records: Array<[bigint, Buffer]>,
  config: WriterConfig,
): Buffer {
  const w = new GolfWriter(config);
  for (const [ts, val] of records) {
    w.append(ts, val);
  }
  return w.seal();
}

describe('Roundtrip', () => {
  test('no compression', () => {
    const records: Array<[bigint, Buffer]> = Array.from({ length: 10 }, (_, i) => [
      BigInt(i) * 1000n,
      makeValue(i, 16),
    ]);
    const data = buildInMemory(records, {
      recordValueSize: 16,
      compression: Compression.None,
      blockCapacity: 4,
    });

    const reader = GolfReader.fromBuffer(data);
    assert.equal(reader.recordCount, 10n);
    assert.deepEqual(reader.timestampRange, [0n, 9000n]);

    const result = reader.query(0n, 9000n);
    assert.equal(result.length, 10);
    for (let i = 0; i < result.length; i++) {
      assert.equal(result[i].timestamp, BigInt(i) * 1000n);
      assert.deepEqual(result[i].value, makeValue(i, 16));
    }
  });

  test('LZ4', () => {
    const records: Array<[bigint, Buffer]> = Array.from({ length: 100 }, (_, i) => [
      BigInt(i) * 500n,
      makeValue(i, 32),
    ]);
    const data = buildInMemory(records, {
      recordValueSize: 32,
      compression: Compression.Lz4,
      blockCapacity: 8,
      tsResolution: TimestampResolution.Microseconds,
    });

    const reader = GolfReader.fromBuffer(data);
    assert.equal(reader.recordCount, 100n);

    const result = reader.query(5000n, 10000n);
    const timestamps = result.map(r => r.timestamp);
    const expected = Array.from({ length: 11 }, (_, i) => BigInt(i + 10) * 500n);
    assert.deepEqual(timestamps, expected);
  });
});

describe('Edge cases', () => {
  test('empty range', () => {
    const records: Array<[bigint, Buffer]> = Array.from({ length: 10 }, (_, i) => [
      BigInt(i) * 100n,
      Buffer.alloc(8),
    ]);
    const data = buildInMemory(records, {
      recordValueSize: 8,
      compression: Compression.None,
      blockCapacity: 4,
    });
    const reader = GolfReader.fromBuffer(data);

    // After all records
    assert.equal(reader.query(1000n, 2000n).length, 0);

    // In gap
    assert.equal(reader.query(50n, 99n).length, 0);
  });

  test('full range', () => {
    const records: Array<[bigint, Buffer]> = Array.from({ length: 20 }, (_, i) => [
      BigInt(i) * 10n,
      Buffer.alloc(8, i & 0xff),
    ]);
    const data = buildInMemory(records, {
      recordValueSize: 8,
      compression: Compression.Lz4,
      blockCapacity: 4,
    });
    const reader = GolfReader.fromBuffer(data);

    const result = reader.query(0n, 0xffffffffffffffffn);
    assert.equal(result.length, 20);
  });

  test('unsorted input', () => {
    const w = new GolfWriter({
      recordValueSize: 4,
      compression: Compression.None,
      blockCapacity: 4,
    });
    const pairs: Array<[bigint, number]> = [
      [300n, 3], [100n, 1], [200n, 2], [500n, 5], [400n, 4],
    ];
    for (const [ts, v] of pairs) {
      w.append(ts, Buffer.from([v, 0, 0, 0]));
    }
    const data = w.seal();
    const reader = GolfReader.fromBuffer(data);

    const result = reader.query(0n, 600n);
    const timestamps = result.map(r => r.timestamp);
    assert.deepEqual(timestamps, [100n, 200n, 300n, 400n, 500n]);
  });

  test('duplicate timestamps', () => {
    const w = new GolfWriter({
      recordValueSize: 4,
      compression: Compression.None,
      blockCapacity: 4,
    });
    for (let i = 0; i < 5; i++) {
      w.append(100n, Buffer.from([i, 0, 0, 0]));
    }
    w.append(200n, Buffer.from([5, 0, 0, 0]));
    const data = w.seal();
    const reader = GolfReader.fromBuffer(data);

    assert.equal(reader.query(100n, 100n).length, 5);
    assert.equal(reader.query(100n, 200n).length, 6);
  });
});

describe('Metadata', () => {
  test('roundtrip', () => {
    const data = buildInMemory(
      Array.from({ length: 5 }, (_, i) => [BigInt(i) * 100n, Buffer.alloc(8)] as [bigint, Buffer]),
      {
        recordValueSize: 8,
        compression: Compression.None,
        blockCapacity: 4,
        metadata: [
          { key: 'source', value: 'test-suite' },
          { key: 'version', value: '1.0.0' },
        ],
      },
    );
    const reader = GolfReader.fromBuffer(data);
    assert.equal(reader.metadata.length, 2);
    assert.equal(reader.metadata[0].key, 'source');
    assert.equal(reader.metadata[0].value, 'test-suite');
  });
});

describe('Cross-language fixtures', () => {
  const fixtureExists = fs.existsSync(TESTDATA_DIR);

  test('read small.golf', { skip: !fixtureExists }, () => {
    const reader = GolfReader.open(path.join(TESTDATA_DIR, 'small.golf'));
    assert.equal(reader.recordCount, 20n);
    assert.equal(reader.header.recordValueSize, 8);

    const result = reader.query(0n, 19000n);
    assert.equal(result.length, 20);
    assert.equal(result[0].timestamp, 0n);
    assert.equal(result[0].value[0], 0);
    assert.equal(result[result.length - 1].timestamp, 19000n);
    assert.equal(result[result.length - 1].value[0], 19);
  });

  test('read with_metadata.golf', { skip: !fixtureExists }, () => {
    const reader = GolfReader.open(path.join(TESTDATA_DIR, 'with_metadata.golf'));
    assert.equal(reader.metadata.length, 2);
    assert.equal(reader.metadata[0].key, 'source');
    assert.equal(reader.metadata[0].value, 'fixture-generator');
  });

  test('read compressed_lz4.golf', { skip: !fixtureExists }, () => {
    const reader = GolfReader.open(path.join(TESTDATA_DIR, 'compressed_lz4.golf'));
    assert.equal(reader.recordCount, 100n);
    const result = reader.query(0n, 49500n);
    assert.equal(result.length, 100);
  });

  // Cross-language: Go fixtures
  test('read go_small.golf', { skip: !fs.existsSync(path.join(TESTDATA_DIR, 'go_small.golf')) }, () => {
    const reader = GolfReader.open(path.join(TESTDATA_DIR, 'go_small.golf'));
    assert.equal(reader.recordCount, 20n);
    const result = reader.query(0n, 19000n);
    assert.equal(result.length, 20);
    assert.equal(result[0].value[0], 0);
    assert.equal(result[19].value[0], 19);
  });

  test('read go_lz4.golf', { skip: !fs.existsSync(path.join(TESTDATA_DIR, 'go_lz4.golf')) }, () => {
    const reader = GolfReader.open(path.join(TESTDATA_DIR, 'go_lz4.golf'));
    assert.equal(reader.recordCount, 100n);
    const result = reader.query(0n, 49500n);
    assert.equal(result.length, 100);
  });

  test('read go_metadata.golf', { skip: !fs.existsSync(path.join(TESTDATA_DIR, 'go_metadata.golf')) }, () => {
    const reader = GolfReader.open(path.join(TESTDATA_DIR, 'go_metadata.golf'));
    assert.equal(reader.metadata[0].key, 'source');
    assert.equal(reader.metadata[0].value, 'go-generator');
  });

  // Cross-language: Python fixtures
  test('read py_small.golf', { skip: !fs.existsSync(path.join(TESTDATA_DIR, 'py_small.golf')) }, () => {
    const reader = GolfReader.open(path.join(TESTDATA_DIR, 'py_small.golf'));
    assert.equal(reader.recordCount, 20n);
    const result = reader.query(0n, 19000n);
    assert.equal(result.length, 20);
  });

  test('read py_lz4.golf', { skip: !fs.existsSync(path.join(TESTDATA_DIR, 'py_lz4.golf')) }, () => {
    const reader = GolfReader.open(path.join(TESTDATA_DIR, 'py_lz4.golf'));
    assert.equal(reader.recordCount, 100n);
    const result = reader.query(0n, 49500n);
    assert.equal(result.length, 100);
  });

  test('read py_metadata.golf', { skip: !fs.existsSync(path.join(TESTDATA_DIR, 'py_metadata.golf')) }, () => {
    const reader = GolfReader.open(path.join(TESTDATA_DIR, 'py_metadata.golf'));
    assert.equal(reader.metadata[0].key, 'source');
    assert.equal(reader.metadata[0].value, 'python-generator');
  });
});

File: go/golf_test.go
package golf

import (
	"bytes"
	"os"
	"path/filepath"
	"testing"
)

func buildInMemory(t *testing.T, records []Record, config WriterConfig) []byte {
	t.Helper()
	w := NewWriter(config)
	for _, r := range records {
		if err := w.Append(r.Timestamp, r.Value); err != nil {
			t.Fatalf("append: %v", err)
		}
	}
	var buf bytes.Buffer
	if err := w.Seal(&buf); err != nil {
		t.Fatalf("seal: %v", err)
	}
	return buf.Bytes()
}

func makeRecords(n int, valueSize int, tsStep uint64) []Record {
	records := make([]Record, n)
	for i := 0; i < n; i++ {
		val := make([]byte, valueSize)
		val[0] = byte(i & 0xff)
		records[i] = Record{
			Timestamp: uint64(i) * tsStep,
			Value:     val,
		}
	}
	return records
}

func TestRoundtripNoCompression(t *testing.T) {
	records := makeRecords(20, 8, 1000)
	data := buildInMemory(t, records, WriterConfig{
		RecordValueSize: 8,
		Compression:     CompressionNone,
		BlockCapacity:   4,
	})

	reader, err := OpenReader(bytes.NewReader(data))
	if err != nil {
		t.Fatalf("open: %v", err)
	}
	if reader.Header.RecordCount != 20 {
		t.Fatalf("expected 20 records, got %d", reader.Header.RecordCount)
	}

	result, err := reader.Query(0, 19000)
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	if len(result) != 20 {
		t.Fatalf("expected 20 results, got %d", len(result))
	}
	for i, r := range result {
		if r.Timestamp != uint64(i)*1000 {
			t.Errorf("record %d: expected ts %d, got %d", i, i*1000, r.Timestamp)
		}
	}
}

func TestRoundtripLz4(t *testing.T) {
	records := makeRecords(100, 16, 500)
	data := buildInMemory(t, records, WriterConfig{
		RecordValueSize: 16,
		Compression:     CompressionLz4,
		BlockCapacity:   8,
	})

	reader, err := OpenReader(bytes.NewReader(data))
	if err != nil {
		t.Fatalf("open: %v", err)
	}

	result, err := reader.Query(5000, 10000)
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	if len(result) != 11 {
		t.Fatalf("expected 11 results, got %d", len(result))
	}
}

func TestRoundtripZstd(t *testing.T) {
	records := makeRecords(50, 8, 100)
	data := buildInMemory(t, records, WriterConfig{
		RecordValueSize: 8,
		Compression:     CompressionZstd,
		BlockCapacity:   16,
	})

	reader, err := OpenReader(bytes.NewReader(data))
	if err != nil {
		t.Fatalf("open: %v", err)
	}

	result, err := reader.Query(2500, 2500)
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	if len(result) != 1 {
		t.Fatalf("expected 1 result, got %d", len(result))
	}
	if result[0].Timestamp != 2500 {
		t.Errorf("expected ts 2500, got %d", result[0].Timestamp)
	}
}

func TestUnsortedInput(t *testing.T) {
	w := NewWriter(WriterConfig{
		RecordValueSize: 4,
		Compression:     CompressionNone,
		BlockCapacity:   4,
	})
	pairs := []struct {
		ts  uint64
		val byte
	}{
		{300, 3}, {100, 1}, {200, 2}, {500, 5}, {400, 4},
	}
	for _, p := range pairs {
		w.Append(p.ts, []byte{p.val, 0, 0, 0})
	}
	var buf bytes.Buffer
	w.Seal(&buf)

	reader, _ := OpenReader(bytes.NewReader(buf.Bytes()))
	result, _ := reader.Query(0, 600)
	expected := []uint64{100, 200, 300, 400, 500}
	for i, r := range result {
		if r.Timestamp != expected[i] {
			t.Errorf("record %d: expected %d, got %d", i, expected[i], r.Timestamp)
		}
	}
}

func TestMetadataRoundtrip(t *testing.T) {
	records := makeRecords(5, 8, 100)
	data := buildInMemory(t, records, WriterConfig{
		RecordValueSize: 8,
		Compression:     CompressionNone,
		BlockCapacity:   4,
		Metadata: []MetadataEntry{
			{Key: "source", Value: "test-suite"},
			{Key: "version", Value: "1.0.0"},
		},
	})

	reader, err := OpenReader(bytes.NewReader(data))
	if err != nil {
		t.Fatalf("open: %v", err)
	}
	if len(reader.Metadata) != 2 {
		t.Fatalf("expected 2 metadata entries, got %d", len(reader.Metadata))
	}
	if reader.Metadata[0].Key != "source" || reader.Metadata[0].Value != "test-suite" {
		t.Errorf("unexpected metadata[0]: %+v", reader.Metadata[0])
	}
}

func TestEmptyRange(t *testing.T) {
	records := makeRecords(10, 8, 100)
	data := buildInMemory(t, records, WriterConfig{
		RecordValueSize: 8,
		Compression:     CompressionNone,
		BlockCapacity:   4,
	})

	reader, _ := OpenReader(bytes.NewReader(data))

	// Query after all records
	result, _ := reader.Query(1000, 2000)
	if len(result) != 0 {
		t.Errorf("expected 0 results, got %d", len(result))
	}
}

// Cross-language fixture tests
func TestReadRustFixtureSmall(t *testing.T) {
	path := filepath.Join("..", "testdata", "small.golf")
	if _, err := os.Stat(path); os.IsNotExist(err) {
		t.Skip("fixture not found, run Rust generate_fixtures first")
	}
	reader, err := OpenFile(path)
	if err != nil {
		t.Fatalf("open fixture: %v", err)
	}
	if reader.Header.RecordCount != 20 {
		t.Fatalf("expected 20 records, got %d", reader.Header.RecordCount)
	}
	if reader.Header.RecordValueSize != 8 {
		t.Fatalf("expected value size 8, got %d", reader.Header.RecordValueSize)
	}
	result, err := reader.Query(0, 19000)
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	if len(result) != 20 {
		t.Fatalf("expected 20 results, got %d", len(result))
	}
	// Verify first record
	if result[0].Timestamp != 0 || result[0].Value[0] != 0 {
		t.Errorf("unexpected first record: ts=%d val[0]=%d", result[0].Timestamp, result[0].Value[0])
	}
	// Verify last record
	last := result[len(result)-1]
	if last.Timestamp != 19000 || last.Value[0] != 19 {
		t.Errorf("unexpected last record: ts=%d val[0]=%d", last.Timestamp, last.Value[0])
	}
}

func TestReadRustFixtureWithMetadata(t *testing.T) {
	path := filepath.Join("..", "testdata", "with_metadata.golf")
	if _, err := os.Stat(path); os.IsNotExist(err) {
		t.Skip("fixture not found")
	}
	reader, err := OpenFile(path)
	if err != nil {
		t.Fatalf("open fixture: %v", err)
	}
	if len(reader.Metadata) != 2 {
		t.Fatalf("expected 2 metadata entries, got %d", len(reader.Metadata))
	}
	if reader.Metadata[0].Key != "source" || reader.Metadata[0].Value != "fixture-generator" {
		t.Errorf("unexpected metadata: %+v", reader.Metadata[0])
	}
}

// ── Cross-language compatibility: read fixtures from Python and TypeScript ──

func assertSmallFixture(t *testing.T, path string) {
	t.Helper()
	if _, err := os.Stat(path); os.IsNotExist(err) {
		t.Skipf("fixture not found: %s", path)
	}
	reader, err := OpenFile(path)
	if err != nil {
		t.Fatalf("open %s: %v", path, err)
	}
	if reader.Header.RecordCount != 20 {
		t.Fatalf("expected 20 records, got %d", reader.Header.RecordCount)
	}
	result, err := reader.Query(0, 19000)
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	if len(result) != 20 {
		t.Fatalf("expected 20 results, got %d", len(result))
	}
	if result[0].Timestamp != 0 || result[0].Value[0] != 0 {
		t.Errorf("unexpected first record")
	}
	last := result[len(result)-1]
	if last.Timestamp != 19000 || last.Value[0] != 19 {
		t.Errorf("unexpected last record")
	}
}

func assertLz4Fixture(t *testing.T, path string) {
	t.Helper()
	if _, err := os.Stat(path); os.IsNotExist(err) {
		t.Skipf("fixture not found: %s", path)
	}
	reader, err := OpenFile(path)
	if err != nil {
		t.Fatalf("open %s: %v", path, err)
	}
	if reader.Header.RecordCount != 100 {
		t.Fatalf("expected 100 records, got %d", reader.Header.RecordCount)
	}
	result, err := reader.Query(0, 49500)
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	if len(result) != 100 {
		t.Fatalf("expected 100 results, got %d", len(result))
	}
}

func assertMetadataFixture(t *testing.T, path, expectedSource string) {
	t.Helper()
	if _, err := os.Stat(path); os.IsNotExist(err) {
		t.Skipf("fixture not found: %s", path)
	}
	reader, err := OpenFile(path)
	if err != nil {
		t.Fatalf("open %s: %v", path, err)
	}
	if len(reader.Metadata) != 2 {
		t.Fatalf("expected 2 metadata entries, got %d", len(reader.Metadata))
	}
	if reader.Metadata[0].Key != "source" || reader.Metadata[0].Value != expectedSource {
		t.Errorf("unexpected metadata: %+v", reader.Metadata[0])
	}
}

func TestReadPythonSmall(t *testing.T) {
	assertSmallFixture(t, filepath.Join("..", "testdata", "py_small.golf"))
}

func TestReadPythonLz4(t *testing.T) {
	assertLz4Fixture(t, filepath.Join("..", "testdata", "py_lz4.golf"))
}

func TestReadPythonMetadata(t *testing.T) {
	assertMetadataFixture(t, filepath.Join("..", "testdata", "py_metadata.golf"), "python-generator")
}

func TestReadTypescriptSmall(t *testing.T) {
	assertSmallFixture(t, filepath.Join("..", "testdata", "ts_small.golf"))
}

func TestReadTypescriptLz4(t *testing.T) {
	assertLz4Fixture(t, filepath.Join("..", "testdata", "ts_lz4.golf"))
}

func TestReadTypescriptMetadata(t *testing.T) {
	assertMetadataFixture(t, filepath.Join("..", "testdata", "ts_metadata.golf"), "typescript-generator")
}
